optimizely/helpers/condition.py

```diff
@@ -12,115 +12,348 @@
 # limitations under the License.
 
 import json
+import math
+import numbers
 
+from six import string_types
 
-class ConditionalOperatorTypes(object):
+
+class ConditionOperatorTypes(object):
   AND = 'and'
   OR = 'or'
   NOT = 'not'
 
 
-DEFAULT_OPERATOR_TYPES = [
-  ConditionalOperatorTypes.AND,
-  ConditionalOperatorTypes.OR,
-  ConditionalOperatorTypes.NOT
-]
-
-
-class ConditionEvaluator(object):
-  """ Class encapsulating methods to be used in audience condition evaluation. """
-
-  def __init__(self, condition_data, attributes):
-    self.condition_data = condition_data
-    self.attributes = attributes
-
-  def evaluator(self, condition):
-    """ Method to compare single audience condition against provided user data i.e. attributes.
-
-    Args:
-      condition: Integer representing the index of condition_data that needs to be used for comparison.
+class ConditionMatchTypes(object):
+  EXACT = 'exact'
+  EXISTS = 'exists'
+  GREATER_THAN = 'gt'
+  LESS_THAN = 'lt'
+  SUBSTRING = 'substring'
 
-    Returns:
-      Boolean indicating the result of comparing the condition value against the user attributes.
-    """
 
-    return self.attributes.get(self.condition_data[condition][0]) == self.condition_data[condition][1]
+class ConditionTreeEvaluator(object):
+  """ Class encapsulating methods to be used in audience condition tree evaluation. """
 
-  def and_evaluator(self, conditions):
+  def and_evaluator(self, conditions, leaf_evaluator):
     """ Evaluates a list of conditions as if the evaluator had been applied
     to each entry and the results AND-ed together
 
     Args:
       conditions: List of conditions ex: [operand_1, operand_2]
+      leaf_evaluator: Function which will be called to evaluate leaf condition values
 
     Returns:
-      Boolean: True if all operands evaluate to True
+      Boolean:
+        - True if all operands evaluate to True
+        - False if a single operand evaluates to False
+      None: if conditions couldn't be evaluated
```

Comment:
Not sure. Would definitely get verbose. Perhaps we could just force ourselves to be diligent in documenting the new, collective behavior somewhere else? E.g. in the KB or at the very top of the condition tree module (rather than separately for each operator's evaluator)